### PR TITLE
Mips port

### PR DESCRIPTION
**Pull request by rbalint**
The attached patches makes XBMC build on mipsel (tested on CI20) and the GUI also runs with the patches back-ported to Gotham.
Currently master can't be started on MIPS due to a crash in the internal FFmpeg copy.

### CODE REVIEW

**Comment by rbalint** [comment]
I managed to run patched Kodi 14.0 using --with-ffmpeg=shared configure option with the ffmpeg packages here:
https://people.debian.org/~rbalint/ppa/xbmc-ffmpeg/xbmc-ffmpeg-wheezy-backports/
I can post a screenshot of the GUI running, but playing video still crashes in the vendor-provided libPVROGL_MESA.so.1.11.2580795 like with XBMC.
I have already reported the issue here:
https://groups.google.com/d/topic/mips-creator-ci20/zZOm38HX81g

**Comment by AndriyP** [comment]
I was little bit late with own pr #5759 :-) But it works fine. While this PR is not merged please include next fix:

Kodi builds libXBMC_addon-mips.so but during loading binary addons expects libXBMC_addon-mipsel-linux.so. And as result Kodi cant load any binary addon. So that names should be synced. Fix is below.

```
--- a/addons/library.xbmc.addon/libXBMC_addon.h
+++ b/addons/library.xbmc.addon/libXBMC_addon.h
@@ -55,8 +55,8 @@ typedef intptr_t      ssize_t;
 #define ADDON_HELPER_ARCH       "powerpc64-linux"
 #elif defined(__ARMEL__)
 #define ADDON_HELPER_ARCH       "arm"
-#elif defined(_MIPSEL)
-#define ADDON_HELPER_ARCH       "mipsel-linux"
+#elif defined(__mips__)
+#define ADDON_HELPER_ARCH       "mips"
 #else
 #define ADDON_HELPER_ARCH       "i486-linux"
 #endif
```

**Comment by AndriyP** [comment]
>> Currently master can't be started on MIPS due to a crash in the internal FFmpeg copy.
Maybe I can help, my build works with internal ffmpeg.

**Comment by rbalint** [comment]
@AndriyP Thanks, I have added your patch to the PR. Do I understand correctly that _this_ PR works for you? In that case @MartijnKaijser could it please be merged?

**Comment by AndriyP** [comment]
To be precisely, I am working with build that doesnt have asm mips atomic operations that you added in this PR. All rest changes are same, I didnt tried to pick up atomic operations because it works without as well. Will integrate them when this PR will be merged.

Regarding FFMPEG for our platform I had to disable some specific MIPS optimization. Here additional ffmpeg configure options added for internal build:
```
--disable-mips32r2 \
--disable-mipsdspr1 \
--disable-mipsdspr2
```
Maybe it make sense to disable them by default. It can be enabled for specific platform with that CPU features on demand.

**Comment by rbalint** [comment]
@AndriyP Most probably those flags made the trick. They are used during the ffmpeg package build as well: https://buildd.debian.org/status/fetch.php?pkg=ffmpeg&arch=mipsel&ver=7%3A2.5.3-1&stamp=1421595399
Since those are platform specific I don't think they should be used in Kodi's build scripts by default.
@fritsch Could this PR be merge since it has been tested by two of us?

File: xbmc/threads/MipsAtomics.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+ *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ *  http://www.gnu.org/copyleft/gpl.html
+ *
+ * Copyright (C) 2003, 06, 07 by Ralf Baechle (ralf@linux-mips.org)
+ *
+ * Most of this file was borrowed from the linux kernel.
+ */
+
+#ifndef __ATOMIC_MIPS_H_
+#define __ATOMIC_MIPS_H_
```
Comment (FernetMenta) [inline xbmc/threads/MipsAtomics.h]: we use pragma once which is supported by all compilers we support

**Comment by AndriyP** [comment]
>> Since those are platform specific I don't think they should be used in Kodi's build scripts by default.
Have same opinion. So maybe it make sense to change this in this PR?

**Comment by rbalint** [comment]
@AndriyP No, I mean we should not add "--disable-mips32r2" etc. , just use the defaults like now.

**Comment by wsnipex** [comment]
please add the needed flags to https://github.com/xbmc/xbmc/blob/master/tools/depends/target/ffmpeg/Makefile
if native building on mips should be supported, then in https://github.com/xbmc/xbmc/blob/master/tools/depends/target/ffmpeg/autobuild.sh as well

**Comment by AndriyP** [comment]
@rbalint I thought that mipsr2 and dsp support is platform specific. And "default" is minimum optimisation to cover more platforms.

**Comment by rbalint** [comment]
Kodi starts and plays video using SW rendering with the patches in the PR:
http://balintreczey.hu/blog/xbmc-from-debian-running-on-mips-ci20-dev-board
@AndriyP I meant we could ask FFmpeg leave mipsr2 and dsp support disabled by default (use GCC's default), but I have added the configure flags here now..

**Comment by rbalint** [comment]
@MartijnKaijser @fritsch @wsnipex @FernetMenta Is there anything missing I should do in this PR?

**Comment by wsnipex** [comment]
its fine from my side. jenkins build this please

**Comment by FernetMenta** [comment]
I am fine with it, thanks

File: xbmc/threads/MipsAtomics.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, write to
+ *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ *  http://www.gnu.org/copyleft/gpl.html
```
Comment (MartijnKaijser) [inline xbmc/threads/MipsAtomics.h]: Didn' t i ask to get this updated first?

**Comment by rbalint** [comment]
@MartijnKaijser sorry fixed the date now
@wsnipex the build seems to be failing due to a problem elsewhere and the problem has been fixed in an other PR already

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
You still have the old header in it. The address of is wrong there. Fix it and squash into the correct commit

**Comment by rbalint** [comment]
@MartijnKaijser done.

**Comment by MartijnKaijser** [comment]
thx
jenkins build this please

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by bkuhls** [comment]
Hi, just a comment: using buildroot with a Sourcery CodeBench toolchain (mips-2014.11-22-mips-linux-gnu-i686-pc-linux-gnu.tar.bz2), current Kodi git master, your patch and a tweaked Kodi buildroot package (to enable mips support) I could compile a mips binary. Due to lack of mips hardware I could not do a run-time test though.

------------------------
  Kodi Configuration:
------------------------
  Kodi Version: 15.0-ALPHA1
  git Rev.:     db54f53b
  Debugging:    No
  Profiling:    No
  Optimization: Yes
  SWIG Available:       Yes
  JRE Available:        Yes
  Doxygen Available:    No
  Crosscomp.:   Yes
  target ARCH:  mips
  target CPU:
  OpenGL:       Yes
  ALSA:         Yes
  DBUS:         No
  VDPAU:        No
  VAAPI:        Yes
  VTBDecoder:   No
  OpenMax:      No
  Joystick:     No
  XRandR:       Yes
  Waveform:     Yes
  Spectrum:     Yes
  GOOM:         Yes
  RSXS:         Yes
  FishBMC:      Yes
  ProjectM:     No
  Touch skin:   No
  X11:          Yes
  Wayland:      No
  Bluray:       Yes
  TexturePacker:Yes
  MID Support:  No
  ccache:       No
  ALSA Support: Yes
  PulseAudio:   No
  DVDCSS:       Yes
  Google Test Framework Configured:     Yes
  Avahi:        No
  mDNSEmbedded: No
  Non-free:     Yes
  ASAP Codec:   No
  MySQL:        No
  Webserver:    Yes
  libssh support:       No
  libRTMP support:      Yes
  libsmbclient support: Yes
  libnfs client support:Yes
  libafpclient support: No
  AirPlay support:      Yes
  AirTunes support (libshairplay):      Yes
  UPnP support:         Yes
  Optical drive:        Yes
  libudev support:      Yes
  libusb support:       No
  libcec support:       Yes
  libcap support:       No
  additional players:   No
  additional codecs:    No
  PVR add-ons:          No
  prefix:       /usr

$ file output/target/usr/lib/kodi/kodi.bin
output/target/usr/lib/kodi/kodi.bin: ELF 32-bit LSB executable, MIPS, MIPS32 rel2 version 1, dynamically linked (uses shared libs), for GNU/Linux 2.6.32, with unknown capability 0xf41 = 0x756e6700, with unknown capability 0x70100 = 0x3040000, not stripped

**Pull request merged**